### PR TITLE
RRF: enable A-Z, Z-A, oldest and newest sort in print menu

### LINKED ISSUE #1981: [FR] Files in icon mode in RRF
With the new soon to be merged thumbnail support, I would like to ask some rework/rethinking of the file menu for in RRF.
Currently only list mode is supported, it can't be turned off/switched to icon mode. 
We discussed earlier this, but maybe this new thumbnail PR could help with speed. Even there would be some latency to load thumbnails, I and possibly many others would use thumbnails.

### PR DESCRIPTION
**Pull request by pfn**
### Requirements

No new requirements

### Description

* Adds `stream_handler` to requestCommandInfo to support >5K payloads
* Adjusts scope of parser/handler in RRFParseACK
* Unlocks icon/grid mode print job listing, fixes #1981
* Resets macro menu cur_page to 0 upon entry (fixes state between print
  and macro menus)
* Code in place to support `M20 S3` once the kinks are worked out

### Benefits

Improves RRF support

### Related Issues

Addresses #1981 request for print menu icon mode

### Testing

Verified working on my MKSTFT28

### PR Status

Ready to merge pending reviews.

### Help Wanted

It seems I've made some memory allocation mistakes when parsing the JSON for `M20 S3` -- it would be greatly appreciated if someone could throw some eyeballs at my changes to see where I goofed? I'm thinking I have some problems around allocating/deallocating `file_name` and `display_name` but can't come to terms with what is happening.

### CODE REVIEW

**Comment by bttguy** [comment]
@pfn Appreciate your contributions! Thank you!

**Comment by pfn** [comment]
Hmm, it looks like it's overflowing user_heap_space on 48K ram devices... I don't know what to do with this...

**Comment by bttguy** [comment]
Which part of/function you added is too large?

**Comment by pfn** [comment]
> Which part of/function you added is too large, and by how much? I mean I see the bytes, but what does it mean? Would it be a sollution to go trough the function of the TFT and disable some of it which is totally or practically useless for RRF users, to enable the RRF related features that would really make a difference for us? What do you think @pfn?

@bttguy I fixed this by removing `M20 S3` support.

**Comment by bttguy** [comment]
> > Which part of/function you added is too large, and by how much? I mean I see the bytes, but what does it mean? Would it be a sollution to go trough the function of the TFT and disable some of it which is totally or practically useless for RRF users, to enable the RRF related features that would really make a difference for us? What do you think @pfn?
> 
> @bttguy I fixed this by removing `M20 S3` support.

I see. And what do you think in general about in the future (with @bigtreetech help), is it possible from your coder's viewpoint to go trough and disable functions so save us RAM, without messing with Marlin users?

**Comment by pfn** [comment]
> I see. So the sorting wont work for now? And what do you think in general about in the future (with @bigtreetech help), is it possible from your coder's viewpoint to go trough and disable functions so save us RAM, without messing with Marlin users?

@bttguy sorting works, but it relies on the implicit oldest-first ordering returned by `M20` -- it's not guaranteed to always be oldest first. Being able to use `M20 S3` means that timestamps can be read and explicitly ordered and there's no reliance on "undocumented behavior"

tl;dr -- it works now, maybe not in the future, depending on how RRF changes.

**Comment by bttguy** [comment]
> > I see. So the sorting wont work for now? And what do you think in general about in the future (with @bigtreetech help), is it possible from your coder's viewpoint to go trough and disable functions so save us RAM, without messing with Marlin users?
> 
> @bttguy sorting works, but it relies on the implicit oldest-first ordering returned by `M20` -- it's not guaranteed to always be oldest first. Being able to use `M20 S3` means that timestamps can be read and explicitly ordered and there's no reliance on "undocumented behavior"
> 
> tl;dr -- it works now, maybe not in the future, depending on how/if RRF changes over time.

Still better than nothing, will do for a while. Thanks for the hard work and bugfix! :) Gonna test soon how reliable it is.

File: TFT/src/User/API/RRFParseACK.cpp [hunk 1]
```diff
@@ -154,11 +150,14 @@ void ParseACKJsonParser::value(const char *value)
 
 void rrfParseACK(const char *data)
 {
+  static JsonStreamingParser parser;
+  static ParseACKJsonParser handler;
```
Comment (digant73) [inline TFT/src/User/API/RRFParseACK.cpp:154]: I see you defined multiple instances for parser and handler. Isn't it possible to share them across more files?

Comment (pfn) [reply TFT/src/User/API/RRFParseACK.cpp:154]: There's really only once instead of each handler in global memory (static) one ParseACKJsonParser and one RRFM20Parser. So there's nothing to share there. At some point, I will probably change all usages of RRM20Parser to be a streaming variant, and then it will be shared instance. (stack vs. heap)

The problem with sharing JsonStreamingParser is that it has to maintain current document state (start/end, etc.). Thinking about it, it doesn't seem like it would be reentrant as there is only 1 active document at any given time, and it's either a something being parsed by parseACK or being passed onto cmd_rev_buf/stream_handler. I'll probably make that change shortly.

**Comment by pfn** [comment]
Looks like the suggestion from @digant73 helped free up a lot of memory to support timestamps, placeholder code for supporting `M20 S3` has been restored.

**Comment by Wombat37** [comment]
I'm really excited about this PR.

Printing from the onboard SD card has been a real pain from the TFT screen because I have to scroll through a 10-page list just to get to the file I've just uploaded!

I requested such a change in Issue #1901 and was told that this was impossible to achieve in the TFT firmware and had to be implemented in the RRF code and I was asked to close the issue!

Thank you so much for finding a possible solution to this - it really is much appreciated!

**Comment by Subcircuit** [comment]
@pfn Hi, are you handling M20 S3?

**Comment by pfn** [comment]
> @pfn Hi, are you handling M20 S3?

@Subcircuit yes, but that code path is currently disabled due to memory issues I can't figure out yet.
